### DISCUSSION TITLE
Concern on ID Column Sequential Uniqueness

### DISCUSSION THREAD

**Discussion by shawnweeks** [pocketbase/pocketbase#2333] [Q&A]
I was exploring PocketBase and opened one of the SQLite files in DBeaver to see what the data model looks like and I'm little concerned on how the id columns are populated. The default value is `'r'||lower(hex(randomblob(7)))` which only gives you 7 bytes of randomness while a UUID would require at least 16 bytes. SQLite is also kinda vague on on how sequentially unique the `randomblob()` function is and their example shows `lower(hex(randomblob(16)))` if your wanting a GUID. Since the ID column only allows 15 characters you can't even populate it with your own UUID which would require 32 characters. I can understand why not using sequences might be useful for migrations but not allowing a proper UUID seems like it will cause problems.

**Comment by shawnweeks** [comment]
It looks like the SQLite default isn't used instead RandomStringWithAlphabet at https://github.com/pocketbase/pocketbase/blob/master/tools/security/random.go is used. However the concern is the same, random number generators aren't generally designed to be unique over time like you'd want in a database primary key. Lots of StackOverflow posts discussing this like https://stackoverflow.com/questions/47410990/random-number-generation-guaranteeing-uniqueness-over-time

**Reply by ganigeorgiev** [reply]
Yes, that's correct. The SQLite random default was added just as fallback in case the user decided to perform direct SQL `INSERT` statements and forgot to specify an ID column.

>  However the concern is the same, random number generators aren't generally designed to be unique over time like you'd want in a database primary key.

I'm not sure that I understand your concerns. The same could be said for UUID4 for example.

We use a cryptographically random string on app level as default that should be enough for most use cases. Even if there is a collision, the worst case scenario is that the record would not be inserted and you'll get an error.

**Also keep in mind that the `id` column is not the "real" primary key that SQLite uses.** You can think of it as a special unique text column.

The real primary key is actually the internal sequential [`rowid` column](https://www.sqlite.org/rowidtable.html) that comes by default with all rowid tables:

> Most tables in a typical SQLite database schema are rowid tables. Rowid tables are distinguished by the fact that they all have a unique, non-NULL, signed 64-bit integer [rowid](https://www.sqlite.org/lang_createtable.html#rowid) **that is used as the access key for the data in the underlying [B-tree](https://www.sqlite.org/fileformat2.html#btree) storage engine.**

**Comment by shawnweeks** [comment]
SQLite uses `rowid` a bit different than a traditional RDBMS would then, normally `rowid` is just the physical address of the row and how indexes and other things reference rows directly. I wouldn't call it the real primary key though since your using `id` for the relationships and that basically makes it the de facto primary key. Best case the current id generator is 62^15 unique values while uuid4 is 256^16. You are correct that the likelihood of collisions is essentially zero, just thought it was odd to use a non-standard id generation scheme

**Comment by plicit** [comment]
Alternatively, it would be nice to expose the rowid using an the integer primary key since that would eliminate the double lookups (text15 -> rowid -> row) and thereby allow more efficient junction/association tables.  UUIDs could be supported through a binary blob column or an association table.  

In any case, thanks for your work, Gani!  I'm having fun playing with pocketbase. :)
